dream-web/view/github.ml

```diff
@@ -0,0 +1,399 @@
+      false jobs
+  in
+  let can_rebuild =
+    List.fold_left
+      (fun accum job_info ->
+        accum
+        ||
+        match job_info.Client.outcome with
+        | Active | NotStarted | Passed -> false
+        | Aborted | Failed _ | Undefined _ -> true)
+      false jobs
+  in
+  let buttons =
+    if can_cancel then
+      [
+        form
+          ~a:[ a_action (hash ^ "/cancel"); a_method `Post ]
+          [
+            Unsafe.data csrf_token;
+            input ~a:[ a_input_type `Submit; a_value "Cancel" ] ();
+          ];
+      ]
+    else if can_rebuild then
+      [
+        form
+          ~a:[ a_action (hash ^ "/rebuild-failed"); a_method `Post ]
+          [
+            Unsafe.data csrf_token;
+            button [ txt "Rebuild Failed" ];
+            input
+              ~a:[ a_name "filter"; a_input_type `Hidden; a_value "failed" ]
+              ();
+          ];
+        form
+          ~a:[ a_action (hash ^ "/rebuild-all"); a_method `Post ]
+          [
+            Unsafe.data csrf_token;
+            button [ txt "Rebuild All" ];
+            input
+              ~a:[ a_name "filter"; a_input_type `Hidden; a_value "none" ]
+              ();
+          ];
+      ]
+    else []
+  in
+  Template.instance ~flash_messages
+    [
+      breadcrumbs
+        [ ("github", "github"); (org, org); (repo, repo) ]
+        (short_hash hash);
+      link_github_refs ~org ~repo refs;
+      link_jobs ~org ~repo ~hash jobs;
+      success_msg;
+      fail_msg;
+      return_link;
+      div buttons;
+    ]
+
+let show_step ~org ~repo ~refs ~hash ~jobs ~variant ~job ~status ~csrf_token
+    ?(flash_messages = []) (data, next) =
+  let header, footer =
+    let can_rebuild = status.Current_rpc.Job.can_rebuild in
+    let buttons =
+      if can_rebuild then
+        Tyxml.Html.
+          [
+            form
+              ~a:[ a_action (variant ^ "/rebuild"); a_method `Post ]
+              [
+                Unsafe.data csrf_token;
+                input ~a:[ a_input_type `Submit; a_value "Rebuild" ] ();
+              ];
+          ]
+      else []
+    in
+    let body =
+      Template.instance ~flash_messages
+        Tyxml.Html.
+          [
+            breadcrumbs
+              [
+                ("github", "github");
+                (org, org);
+                (repo, repo);
+                (short_hash hash, "commit/" ^ hash);
+              ]
+              variant;
+            link_github_refs ~org ~repo refs;
+            link_jobs ~org ~repo ~hash ~selected:variant jobs;
+            div buttons;
+            pre [ txt "@@@" ];
+          ]
+    in
+    Astring.String.cut ~sep:"@@@" body |> Option.get
+  in
+  let ansi = Ansi.create () in
+  let open Lwt.Infix in
+  Dream.stream
+    ~headers:[ ("Content-type", "text/html; charset=utf-8") ]
+    (fun response_stream ->
+      let%lwt () = Dream.write response_stream header in
+      let%lwt () = Dream.write response_stream (Ansi.process ansi data) in
```

Comment:
```suggestion
      Dream.write response_stream header >>= fun () ->
      Dream.write response_stream (Ansi.process ansi data) >>= fun () ->
```
Maybe the version without ppx is better?

Reply:
I find the version with the ppx better, it correctly propagates backtraces which is super for debugging. It's also possible to write `;%lwt` for sequences but the readability of _that_ is discutable.

Reply:
Thanks all for your input - I'm switching everything to the version without ppx for consistency.